zero/src/block_interval.rs

```diff
@@ -31,75 +29,51 @@ pub enum BlockInterval {
 }
 
 impl BlockInterval {
-    /// Create a new block interval
+    /// Creates a new block interval.
     ///
-    /// A valid block range is of the form:
-    ///     * `block_number` for a single block number
-    ///     * `lhs..rhs`, `lhs..=rhs` as an exclusive/inclusive range
-    ///     * `lhs..` for a range starting from `lhs` to the chain tip. `lhs..=`
-    ///       is also valid format.
+    /// If end_block is None, the interval is unbounded and will follow from
+    /// start_block. If start_block == end_block, the interval is a single
+    /// block. Otherwise, the interval is an inclusive range from start_block to
+    /// end_block.
     ///
-    /// # Example
-    ///
-    /// ```rust
-    ///    # use alloy::rpc::types::eth::BlockId;
-    ///    # use zero::block_interval::BlockInterval;
-    ///    assert_eq!(BlockInterval::new("0..10").unwrap(), BlockInterval::Range(0..10));
-    ///    assert_eq!(BlockInterval::new("0..=10").unwrap(), BlockInterval::Range(0..11));
-    ///    assert_eq!(BlockInterval::new("32141").unwrap(), BlockInterval::SingleBlockId(BlockId::Number(32141.into())));
-    ///    assert_eq!(BlockInterval::new("100..").unwrap(), BlockInterval::FollowFrom{start_block: 100});
-    /// ```
-    pub fn new(s: &str) -> anyhow::Result<BlockInterval> {
-        if (s.starts_with("0x") && s.len() == 66) || s.len() == 64 {
-            // Try to parse hash
-            let hash = s
-                .parse::<B256>()
-                .map_err(|_| anyhow!("invalid block hash '{s}'"))?;
-            return Ok(BlockInterval::SingleBlockId(BlockId::Hash(hash.into())));
-        }
-
-        // First we parse for inclusive range and then for exclusive range,
-        // because both separators start with `..`
-        if let Ok(range) = parsing::parse_range_inclusive(s) {
-            Ok(BlockInterval::Range(range))
-        } else if let Ok(range) = parsing::parse_range_exclusive(s) {
-            Ok(BlockInterval::Range(range))
-        }
-        // Now we look for the follow from range
-        else if s.contains("..") {
-            let mut split = s.trim().split("..").filter(|s| *s != "=" && !s.is_empty());
+    /// end_block is always treated as inclusive because it may have been
+    /// specified as a block hash.
+    pub async fn new(
+        provider: Arc<impl ZeroBlockProvider>,
```

Comment:
smell: doesn't feel right to need an Arc here

Reply:
No smell here. We use one provider in the whole program, we pass it from the top level using Arc.

Reply:
I think this is a case of [needless pass by value](https://rust-lang.github.io/rust-clippy/master/index.html#needless_pass_by_value)

- ZeroBlockProvider only has `&self` receivers
- So the function only needs `[&] impl ZeroBlockProvider` (note that `alloy_provider::Provider` already does the `impl<T: Provider> Provider for &T { .. }`

The function should require as little from its callers as possible

Reply:
@0xaatif In the leader we have couple of threads using provider. We only instantiate one instance at the start of the program and pass it from the top and multiple async tasks use this provider in parallel. I agree that in this particular case (in particular where the block interval is created in the leader) pass by value (or with Arc smart pointer) is not needed, but it was easier and cleaner to by convention maintain the same Provider argument through the whole application than to create Provider, pass it by reference and then wrap it in Arc to pass it further to other tasks.  Cloning the `Arc` is just increasing the reference counter so it is not really an overhead.

Reply:
P.S. Earlier, Provider was indeed created and handled as an common reference. But during refactors it often happen that functions were called from other tasks, so the only way they could get the provider is through Arc pointer. Hence the Arc<Provider> convention was born.